File: src-tauri/src/utils/process.rs
use std::path::{Path, PathBuf};

#[cfg(target_os = "windows")]
use std::os::windows::process::CommandExt;

#[cfg(target_os = "windows")]
const CREATE_NO_WINDOW: u32 = 0x08000000;

#[cfg(target_os = "windows")]
pub fn windows_hidden_creation_flags() -> u32 {
    CREATE_NO_WINDOW
}

pub fn configure_std_command(command: &mut std::process::Command) -> &mut std::process::Command {
    #[cfg(target_os = "windows")]
    {
        command.creation_flags(CREATE_NO_WINDOW);
    }

    command
}

pub fn configure_tokio_command(
    command: &mut tokio::process::Command,
) -> &mut tokio::process::Command {
    #[cfg(target_os = "windows")]
    {
        command.creation_flags(CREATE_NO_WINDOW);
    }

    command
}

pub fn std_command(program: impl AsRef<std::ffi::OsStr>) -> std::process::Command {
    let mut command = std::process::Command::new(program);
    configure_std_command(&mut command);
    command
}

pub fn tokio_command(program: impl AsRef<std::ffi::OsStr>) -> tokio::process::Command {
    let mut command = tokio::process::Command::new(program);
    configure_tokio_command(&mut command);
    command
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedCommand {
    pub program: String,
    pub args: Vec<String>,
}

pub fn parse_command_string(input: &str) -> Result<ParsedCommand, String> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut quote: Option<char> = None;
    let mut escaped = false;

    for ch in input.trim().chars() {
        if escaped {
            current.push(ch);
            escaped = false;
            continue;
        }

        match ch {
            '\\' if quote.is_some() => escaped = true,
            '"' | '\'' => {
                if quote == Some(ch) {
                    quote = None;
                } else if quote.is_none() {
                    quote = Some(ch);
                } else {
                    current.push(ch);
                }
            }
            c if c.is_whitespace() && quote.is_none() => {
                if !current.is_empty() {
                    parts.push(std::mem::take(&mut current));
                }
            }
            _ => current.push(ch),
        }
    }

    if escaped {
        current.push('\\');
    }

    if quote.is_some() {
        return Err("Command contains an unmatched quote".to_string());
    }

    if !current.is_empty() {
        parts.push(current);
    }

    let Some(program) = parts.first().cloned() else {
        return Err("CLI command is empty".to_string());
    };

    Ok(ParsedCommand {
        program,
        args: parts.into_iter().skip(1).collect(),
    })
}

pub fn command_exists(cmd: &str) -> bool {
    resolve_command_path(cmd).is_some()
}

pub fn resolve_command_path(cmd: &str) -> Option<PathBuf> {
    let trimmed = cmd.trim();
    if trimmed.is_empty() {
        return None;
    }

    if looks_like_path(trimmed) {
        return resolve_explicit_path(Path::new(trimmed));
    }

    lookup_command_candidates(trimmed).into_iter().next()
}

pub fn lookup_command_candidates(cmd: &str) -> Vec<PathBuf> {
    let mut paths = Vec::new();

    #[cfg(target_os = "windows")]
    {
        let output = std_command("where").arg(cmd).output().ok();
        append_command_output_paths(&mut paths, output);
        append_windows_fallback_paths(&mut paths, cmd);
    }

    #[cfg(not(target_os = "windows"))]
    {
        let output = std_command("which").arg(cmd).output().ok();
        append_command_output_paths(&mut paths, output);
    }

    dedupe_paths(paths)
}

pub fn open_browser_url(url: &str) -> anyhow::Result<()> {
    #[cfg(target_os = "macos")]
    {
        std_command("open")
            .arg(url)
            .spawn()
            .map_err(|e| anyhow::anyhow!("Failed to open browser: {}", e))?;
    }

    #[cfg(target_os = "windows")]
    {
        std_command("rundll32")
            .args(["url.dll,FileProtocolHandler", url])
            .spawn()
            .map_err(|e| anyhow::anyhow!("Failed to open browser: {}", e))?;
    }

    #[cfg(target_os = "linux")]
    {
        std_command("xdg-open")
            .arg(url)
            .spawn()
            .map_err(|e| anyhow::anyhow!("Failed to open browser: {}", e))?;
    }

    Ok(())
}

fn looks_like_path(cmd: &str) -> bool {
    cmd.contains(std::path::MAIN_SEPARATOR)
        || cmd.contains('/')
        || cmd.contains('\\')
        || Path::new(cmd).is_absolute()
}

fn resolve_explicit_path(path: &Path) -> Option<PathBuf> {
    if path.exists() {
        return Some(path.to_path_buf());
    }

    #[cfg(target_os = "windows")]
    {
        for ext in ["exe", "cmd", "bat", "ps1"] {
            let candidate = if path.extension().is_some() {
                path.to_path_buf()
            } else {
                path.with_extension(ext)
            };
            if candidate.exists() {
                return Some(candidate);
            }
        }
    }

    None
}

fn append_command_output_paths(
    paths: &mut Vec<PathBuf>,
    output: Option<std::process::Output>,
) {
    if let Some(output) = output {
        if output.status.success() {
            let stdout = String::from_utf8_lossy(&output.stdout);
            for line in stdout.lines().map(str::trim).filter(|line| !line.is_empty()) {
                paths.push(PathBuf::from(line));
            }
        }
    }
}

#[cfg(target_os = "windows")]
fn append_windows_fallback_paths(paths: &mut Vec<PathBuf>, cmd: &str) {
    let mut candidates = Vec::new();

    if let Ok(app_data) = std::env::var("APPDATA") {
        let base = PathBuf::from(app_data).join("npm");
        candidates.push(base.join(cmd));
        candidates.push(base.join(format!("{cmd}.cmd")));
        candidates.push(base.join(format!("{cmd}.exe")));
        candidates.push(base.join(format!("{cmd}.ps1")));
    }

    if let Ok(local_app_data) = std::env::var("LOCALAPPDATA") {
        let base = PathBuf::from(&local_app_data);
        let npm = base.join("npm");
        candidates.push(npm.join(cmd));
        candidates.push(npm.join(format!("{cmd}.cmd")));
        candidates.push(npm.join(format!("{cmd}.exe")));
        candidates.push(
            base.join("Microsoft")
                .join("WinGet")
                .join("Links")
                .join(format!("{cmd}.exe")),
        );
        candidates.push(base.join("Microsoft").join("WindowsApps").join(format!("{cmd}.exe")));

        for pattern in [
            format!("{}\\Programs\\Python\\Python*\\Scripts\\{cmd}.exe", base.display()),
            format!("{}\\Programs\\Python\\Python*\\Scripts\\{cmd}", base.display()),
        ] {
            if let Ok(entries) = glob::glob(&pattern) {
                for entry in entries.flatten() {
                    candidates.push(entry);
                }
            }
        }
    }

    if let Ok(user_profile) = std::env::var("USERPROFILE") {
        let base = PathBuf::from(user_profile);
        candidates.push(base.join(".local").join("bin").join(format!("{cmd}.exe")));
        candidates.push(base.join(".local").join("bin").join(cmd));
    }

    for candidate in candidates {
        if candidate.exists() {
            paths.push(candidate);
        }
    }
}

fn dedupe_paths(paths: Vec<PathBuf>) -> Vec<PathBuf> {
    let mut seen = std::collections::HashSet::new();
    let mut deduped = Vec::new();

    for path in paths {
        let key = path.to_string_lossy().to_string();
        #[cfg(target_os = "windows")]
        let key = key.to_lowercase();

        if seen.insert(key) {
            deduped.push(path);
        }
    }

    deduped
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_command_string_basic() {
        let parsed = parse_command_string("codex login status").unwrap();
        assert_eq!(parsed.program, "codex");
        assert_eq!(parsed.args, vec!["login", "status"]);
    }

    #[test]
    fn test_parse_command_string_with_quotes() {
        let parsed = parse_command_string("\"C:\\\\Program Files\\\\Codex\\\\codex.exe\" --profile default").unwrap();
        assert_eq!(parsed.program, "C:\\Program Files\\Codex\\codex.exe");
        assert_eq!(parsed.args, vec!["--profile", "default"]);
    }

    #[cfg(target_os = "windows")]
    #[test]
    fn test_windows_hidden_creation_flags_constant() {
        assert_eq!(windows_hidden_creation_flags(), 0x08000000);
    }
}

File: server/utils/process.js
import { execSync, execFileSync } from 'child_process';
import path from 'path';
import fs from 'fs';
import os from 'os';

/**
 * Parse a command string into program + args, handling quotes and escapes.
 * Port of Rust parse_command_string.
 */
export function parseCommandString(input) {
  const trimmed = input.trim();
  const parts = [];
  let current = '';
  let quote = null;
  let escaped = false;

  for (const ch of trimmed) {
    if (escaped) {
      current += ch;
      escaped = false;
      continue;
    }
    if (ch === '\\' && quote) {
      escaped = true;
      continue;
    }
    if (ch === '"' || ch === "'") {
      if (quote === ch) {
        quote = null;
      } else if (!quote) {
        quote = ch;
      } else {
        current += ch;
      }
      continue;
    }
    if (/\s/.test(ch) && !quote) {
      if (current) {
        parts.push(current);
        current = '';
      }
      continue;
    }
    current += ch;
  }

  if (escaped) current += '\\';
  if (quote) throw new Error('Command contains an unmatched quote');
  if (current) parts.push(current);
  if (parts.length === 0) throw new Error('CLI command is empty');

  return { program: parts[0], args: parts.slice(1) };
}

/**
 * Check if a command exists in the system PATH or common fallback locations.
 */
export function commandExists(cmd) {
  return !!resolveCommandPath(cmd);
}

/**
 * Resolve the full path to a command.
 */
export function resolveCommandPath(cmd) {
  const trimmed = (cmd || '').trim();
  if (!trimmed) return null;

  // If it looks like a path (contains separator), check directly
  if (trimmed.includes('/') || trimmed.includes('\\') || path.isAbsolute(trimmed)) {
    if (fs.existsSync(trimmed)) return trimmed;
    if (os.platform() === 'win32') {
      for (const ext of ['exe', 'cmd', 'bat', 'ps1']) {
        const candidate = path.extname(trimmed) ? trimmed : `${trimmed}.${ext}`;
        if (fs.existsSync(candidate)) return candidate;
      }
    }
    return null;
  }

  // Use which/where to find in PATH
  const candidates = lookupCommandCandidates(trimmed);
  return candidates.length > 0 ? candidates[0] : null;
}

/**
 * Look up all candidate paths for a command.
 */
export function lookupCommandCandidates(cmd) {
  const paths = [];

  try {
    if (os.platform() === 'win32') {
      const output = execFileSync('where', [cmd], { encoding: 'utf-8', timeout: 5000 });
      for (const line of output.split('\n').map(l => l.trim()).filter(Boolean)) {
        paths.push(line);
      }
    } else {
      const output = execFileSync('which', [cmd], { encoding: 'utf-8', timeout: 5000 });
      for (const line of output.split('\n').map(l => l.trim()).filter(Boolean)) {
        paths.push(line);
      }
    }
  } catch {
    // Command not found
  }

  // Windows fallback paths
  if (os.platform() === 'win32') {
    const appData = process.env.APPDATA;
    const localAppData = process.env.LOCALAPPDATA;
    const userProfile = process.env.USERPROFILE;

    const candidates = [];
    if (appData) {
      const npmDir = path.join(appData, 'npm');
      candidates.push(path.join(npmDir, cmd), path.join(npmDir, `${cmd}.cmd`), path.join(npmDir, `${cmd}.exe`));
    }
    if (localAppData) {
      const npmDir = path.join(localAppData, 'npm');
      candidates.push(path.join(npmDir, cmd), path.join(npmDir, `${cmd}.cmd`), path.join(npmDir, `${cmd}.exe`));
      candidates.push(path.join(localAppData, 'Microsoft', 'WinGet', 'Links', `${cmd}.exe`));
      candidates.push(path.join(localAppData, 'Microsoft', 'WindowsApps', `${cmd}.exe`));
    }
    if (userProfile) {
      candidates.push(path.join(userProfile, '.local', 'bin', `${cmd}.exe`));
      candidates.push(path.join(userProfile, '.local', 'bin', cmd));
    }
    for (const c of candidates) {
      if (fs.existsSync(c) && !paths.includes(c)) {
        paths.push(c);
      }
    }
  }

  // Deduplicate
  const seen = new Set();
  return paths.filter(p => {
    const key = os.platform() === 'win32' ? p.toLowerCase() : p;
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

/**
 * Open a URL in the default browser.
 */
export function openBrowserUrl(url) {
  try {
    const platform = os.platform();
    if (platform === 'darwin') {
      execSync(`open "${url}"`);
    } else if (platform === 'win32') {
      execSync(`start "" "${url}"`);
    } else {
      execSync(`xdg-open "${url}"`);
    }
  } catch (e) {
    console.error(`Failed to open browser: ${e.message}`);
  }
}
